src/Illuminate/Auth/EloquentUserProvider.php

```diff
@@ -152,7 +152,38 @@ public function validateCredentials(UserContract $user, array $credentials)
             return false;
         }
 
-        return $this->hasher->check($plain, $user->getAuthPassword());
+        if (! $this->hasher->check($plain, $hash = $user->getAuthPassword())) {
+            return false;
+        }
+
+        $this->rehashIfRequired($user, $hash, $plain);
+
+        return true;
+    }
+
+    /**
+     * Rehash the user's password if required.
+     *
+     * @param  \Illuminate\Contracts\Auth\Authenticatable  $user
+     * @param  string  $hash
+     * @param  string  $plain
+     * @return void
+     */
+    protected function rehashIfRequired(UserContract $user, string $hash, string $plain)
+    {
+        $attribute = method_exists($user, 'getAuthPasswordName') ? $user->getAuthPasswordName() : 'password';
+
+        if (! isset($user->{$attribute}) || ! hash_equals($hash, $user->{$attribute})) {
+            return;
+        }
+
+        if (! $this->hasher->needsRehash($hash)) {
+            return;
+        }
+
+        $user->forceFill([
+            $attribute => $this->hasher->make($plain),
```

Comment:
Would this not be a BC for those of us who use an `Attribute` mutator to set and hash the password?

```php
protected function password(): Attribute
{
    return new Attribute(
        set: fn (string $value): string => Hash::make($value),
    );
}
```

Reply:
Yeah, it probably would. 

The built-in hash cast checks if it's hashed before hashing:

```
return $value !== null && ! Hash::isHashed($value) ? Hash::make($value) : $value;
```

But a custom cast like your example wouldn't do this. 🤔